### PR TITLE
[openvpn] Improve parse_data to raise exception if something goes wrong #58

### PR DESCRIPTION
**Pull request by EdgeKing810**
GCI Task: https://codein.withgoogle.com/dashboard/task-instances/5302984465448960/

(P.S. It's now that I'm working on the tests, need some help from Travis)

### CODE REVIEW

**Comment by EdgeKing810** [comment]
Idk if I did it well... I based myself on some code from ``netdiff`` and from ``https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/config/tests/test_config.py#L22-L28``.

File: tests/test_openvpn.py [hunk 1]
```diff
@@ -18,6 +19,15 @@ def test_parse(self):
         self.assertEqual(p.version, '1')
         self.assertEqual(p.metric, 'static')
 
+    def test_to_python(self):
```
Comment (nemesisdesign) [inline tests/test_openvpn.py]: remove this test, it doesn't make sense

**Comment by EdgeKing810** [comment]
Errors and errors... :frowning_face:

**Comment by EdgeKing810** [comment]
Travis build is fixed now... I'll need a review on what to improve or do more :/

File: netdiff/parsers/openvpn.py [hunk 2]
```diff
@@ -16,8 +17,12 @@ class OpenvpnParser(BaseParser):
     def to_python(self, data):
         try:
             return parse_status(data)
-        except (AttributeError, ParsingError):
-            return None
+        except (AttributeError, ParsingError) as e:
+            msg = 'something went wrong: {0}'.format(str(e))
+            try:
```
Comment (nemesisdesign) [inline netdiff/parsers/openvpn.py]: why you can't just remove this `try` block with the following?

```python
raise ConversionException(msg)
```

File: setup.cfg [hunk 3]
```diff
@@ -4,6 +4,7 @@ universal=1
 [flake8]
 exclude = setup.py
 max-line-length = 110
+ignore=F841
```
Comment (nemesisdesign) [inline setup.cfg]: definitely not, please remove this

File: tests/test_openvpn.py [hunk 4]
```diff
@@ -67,6 +67,12 @@ def test_json_dict_tap(self):
     def test_empty_string(self):
         OpenvpnParser(data='{}')
 
+    def test_random_string(self):
+        OpenvpnParser(data='{string123}')
+
+    def test_bogus_data(self):
+        OpenvpnParser(data='{%$^*([[zsde4323@#}')
```
Comment (nemesisdesign) [inline tests/test_openvpn.py]: we expect an exception to be raise here, this test cannot pass like that, you have to do something like here:
https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/config/tests/test_config.py#L22-L28

**Note** I already explained you how to do this via chat. I hope that with this more explicit comment you now understand.

File: netdiff/parsers/openvpn.py (hunk 2 above)
Comment (EdgeKing810) [reply netdiff/parsers/openvpn.py]: because another exception is being raised.

**Comment by EdgeKing810** [comment]
@nemesisdesign I'm getting a ``KeyError`` exception. That's why I was using that ``try - catch`` block. :/

File: netdiff/parsers/openvpn.py [hunk 5]
```diff
@@ -16,8 +17,9 @@ class OpenvpnParser(BaseParser):
     def to_python(self, data):
         try:
             return parse_status(data)
-        except (AttributeError, ParsingError):
-            return None
+        except (AttributeError, ParsingError) as e:
+            msg = 'something went wrong: {0}'.format(str(e))
+            raise ConversionException(msg)
```
Comment (nemesisdesign) [inline netdiff/parsers/openvpn.py]: try changing this to:

```python
raise ConversionException(msg, data=data)
```

**Pull request merged**